File: app/src/main/java/com/vo1d/schedulemanager/v2/ui/instructors/list/InstructorKeyProvider.java
package com.vo1d.schedulemanager.v2.ui.instructors.list;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.selection.ItemKeyProvider;
import androidx.recyclerview.widget.RecyclerView;

class InstructorKeyProvider extends ItemKeyProvider<Long> {

    private final InstructorsListAdapter adapter;
    private final RecyclerView recyclerView;

    public InstructorKeyProvider(InstructorsListAdapter adapter, RecyclerView recyclerView) {
        super(SCOPE_MAPPED);
        this.adapter = adapter;
        this.recyclerView = recyclerView;
    }

    @Nullable
    @Override
    public Long getKey(int position) {
        return adapter.getItemId(position);
    }

    @Override
    public int getPosition(@NonNull Long key) {
        RecyclerView.ViewHolder viewHolder = recyclerView.findViewHolderForItemId(key);
        return viewHolder == null ? RecyclerView.NO_POSITION : viewHolder.getAdapterPosition();
    }
}

File: app/src/main/kotlin/ru/vo1d/timetablemanager/ui/sections/instructors/list/InstructorKeyProvider.kt
package ru.vo1d.timetablemanager.ui.sections.instructors.list

import androidx.recyclerview.selection.ItemKeyProvider
import androidx.recyclerview.widget.RecyclerView

internal class InstructorKeyProvider(
    private val adapter: InstructorsListAdapter,
    private val recyclerView: RecyclerView
) : ItemKeyProvider<Long>(SCOPE_MAPPED) {

    override fun getKey(position: Int) =
        adapter.getItemId(position)

    override fun getPosition(key: Long): Int {
        val viewHolder = recyclerView.findViewHolderForItemId(key)
        return viewHolder?.adapterPosition ?: RecyclerView.NO_POSITION
    }
}